core/src/main/java/org/jdbi/v3/core/cache/internal/InternalJdbiCache.java

```diff
@@ -0,0 +1,167 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jdbi.v3.core.cache.internal;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ConcurrentMap;
+
+import org.jdbi.v3.core.cache.JdbiCache;
+import org.jdbi.v3.core.cache.JdbiCacheBuilder;
+import org.jdbi.v3.core.cache.JdbiCacheLoader;
+
+public final class InternalJdbiCache<K, V> implements JdbiCache<K, V> {
+
+    private final ConcurrentMap<K, V> cache;
+    private final ConcurrentLinkedQueue<K> expungeQueue;
+    private final JdbiCacheLoader<K, V> cacheLoader;
+
+    private final int maxSize;
+
+    private InternalJdbiCache(Builder builder, JdbiCacheLoader<K, V> cacheLoader) {
+        this.cache = new ConcurrentHashMap<>();
+        this.expungeQueue = new ConcurrentLinkedQueue<>();
+        this.cacheLoader = cacheLoader;
+
+        this.maxSize = builder.maxSize;
+    }
+
+    @Override
+    public V get(K key) {
+        try {
+            if (cacheLoader == null) {
+                return cache.get(key);
+            } else {
+                return cache.computeIfAbsent(key, compute(cacheLoader)::create);
+            }
+        } finally {
+            expunge();
```

Comment:
It would be nice to only run `expunge` when a new mapping was added, rather than on every read.

Reply:
as discussed, this would require modification of the map from the loader which is discouraged / has bugs. I made the existing mechanism more efficient.

Reply:
fwiw, this should perform similarly to a synchronized `LinkedHashMap` now, since every read/write must lock due to performing the size check under a synchronized block. The main benefit is that the computation's lock is (roughly) per-entry rather than per-map, so it supports more efficient concurrent loads. There are lots of tweak and approaches you could take to make that better, etc. but that might not be important if this is just to avoid having an explicit dependency (but recommended / implicit by what would be needed in production). Or if this cache isn't under much lock contention then it would also be adequate. But if this is meant to be very fast where fifo/lru are adequate for the needs, then you might consider further tweaks.